### ISSUE TITLE
cipher.update returns null when the input is larger than blocker cipher

### ISSUE THREAD

**Issue by chenzhihao** [aws/amazon-s3-encryption-client-java#185]
### Problem:

Hi there, I've found the issue that `CipherSubscriber.java` try to wrap the null value:

```
wrappedSubscriber.onNext(ByteBuffer.wrap(outputBuffer));
```
The problem is caused by ` outputBuffer = cipher.update(buf, 0, amountToReadFromByteBuffer);` gets the null value from `outputBuffer` even when `amountToReadFromByteBuffer` is larger than cipherBlocker.

https://github.com/aws/amazon-s3-encryption-client-java/blob/main/src/main/java/software/amazon/encryption/s3/internal/CipherSubscriber.java#L52

It can be reproduced by using an AES-128 Key as Encryption material and uploading a random file. I used `dd if=/dev/random of=somefile-100m bs=1M count=100` to generate a random file for testing.

I've added some log and got these information if that helps:
```java
 @Override
    public void onNext(ByteBuffer byteBuffer) {
        int amountToReadFromByteBuffer = getAmountToReadFromByteBuffer(byteBuffer);

        if (amountToReadFromByteBuffer > 0) {
            byte[] buf = BinaryUtils.copyBytesFrom(byteBuffer, amountToReadFromByteBuffer);
            outputBuffer = cipher.update(buf, 0, amountToReadFromByteBuffer);
            if (outputBuffer == null && amountToReadFromByteBuffer < cipher.getBlockSize()) {
                // The underlying data is too short to fill in the block cipher
                // This is true at the end of the file, so complete to get the final
                // bytes
                this.onComplete();
            }

////no output from cipher, where the error happens
            if (outputBuffer == null && amountToReadFromByteBuffer > cipher.getBlockSize()) {
                LogFactory.getLog(getClass()).error("Cipher outputBuffer is null. amountToReadFromByteBuffer: " + amountToReadFromByteBuffer +
                    " cipher.getBlockSize(): " + cipher.getBlockSize() +
                     " byteBuffer: " + byteBuffer +
                    " contentRead: " + contentRead.get() + " buf: " + BinaryUtils.toHex(buf)
                    );
   
            } else {
                wrappedSubscriber.onNext(ByteBuffer.wrap(outputBuffer));
            }
        } else {
            // Do nothing
            wrappedSubscriber.onNext(byteBuffer);
        }
    }



//error log:
//Cipher outputBuffer is null. amountToReadFromByteBuffer: 168 cipher.getBlockSize(): 16 byteBuffer: java.nio.HeapByteBuffer[pos=0 lim=168 cap=168] contentRead: 9593000
```

What I'm not sure about is it does not always happen. I also used some other real files from my local machine, it didn't have this issue. I suspect this is related to the content of the file.

### Solution:
Not sure yet. Can I please get some suggestions from you as I'm not familiar with how the AES cipher works under the hood?

### Out of scope:

Is there anything the solution will intentionally NOT address?

[//]: #  (NOTE: If you believe this might be a security issue, please email aws-security@amazon.com instead of creating a GitHub issue. For more details, see the AWS Vulnerability Reporting Guide: https://aws.amazon.com/security/vulnerability-reporting/ )

**Comment by justplaz** [comment]
Hello, thanks for reaching out! 

You are correct that there is a bug in the `CipherSubscriber` involving `null` values of `outputBuffer`. Please see https://github.com/aws/amazon-s3-encryption-client-java/pull/209 for more details. 

Please also note that in general, the S3EC does not support decryption of data encrypted by another library, and the S3EC does not support content encryption using a 128 bit AES data key. 

The PR above should fix the `NullPointerException` issue. Once it is merged and released, I will update this issue accordingly. Thanks again!

**Comment by imabhichow** [comment]
Hello @chenzhihao,

Thank you again for your report.

We have published a fix in new version [3.1.2](https://github.com/aws/amazon-s3-encryption-client-java/releases/tag/v3.1.2).

Thank you!
